src/main/io/rcdevice_osd.c

```diff
@@ -0,0 +1,241 @@
+        if (video_system == VIDEO_SYSTEM_NTSC) {
+            tvMode = 0;
+        } else if (video_system == VIDEO_SYSTEM_PAL) {
+            tvMode = 1;
+        }
+
+        if (!runcamDeviceWriteSetting(osdDevice, RCDEVICE_PROTOCOL_SETTINGID_DISP_TV_MODE, &tvMode, sizeof(uint8_t), &response)) {
+            return false;
+        }
+
+        if (response.resultCode) {
+            return false;
+        }
+    }
+
+    // user bf charset
+    uint8_t charsetID = 0;
+    runcamDeviceWriteSettingResponse_t updateCharsetResp;
+    if (!runcamDeviceWriteSetting(osdDevice, RCDEVICE_PROTOCOL_SETTINGID_DISP_CHARSET, &charsetID, sizeof(uint8_t), &updateCharsetResp) || updateCharsetResp.resultCode != 0) {
+        return false;
+    }
+
+#ifdef USE_PARTICLE_DRAW
+    memset(shadowBuffer, 2, VIDEO_BUFFER_CHARS_PAL);
+#endif
+
+    // fill screen with ' '
+    rcdeviceOSDClearScreen(NULL);
+    return true;
+}
+
+int rcdeviceOSDGrab(displayPort_t *displayPort)
+{
+    UNUSED(displayPort);
+    // osdResetAlarms();
+    // resumeRefreshAt = 0;
+    return 0;
+}
+
+int rcdeviceOSDRelease(displayPort_t *displayPort)
+{
+    UNUSED(displayPort);
+    return 0;
+}
+
+int rcdeviceOSDDrawScreen(displayPort_t *displayPort)
+{
+    UNUSED(displayPort);
+
+#ifdef USE_PARTICLE_DRAW
+    static uint16_t pos = 0;
+    int k = 0;
+
+    if (!rcdeviceOSDLock) {
+        rcdeviceOSDLock = true;
+
+        uint8_t data[60];
+        uint8_t dataLen = 0;
+        for (k = 0; k < MAX_CHARS2UPDATE; k++) {
+            if (screenBuffer[pos] != shadowBuffer[pos]) {
+                shadowBuffer[pos] = screenBuffer[pos];
+                uint8_t x = pos % columnCount;
+                uint8_t y = pos / columnCount;
+                data[dataLen++] = x;
+                data[dataLen++] = y;
+                data[dataLen++] = screenBuffer[pos];
+            }
+
+            if (++pos >= maxScreenSize) {
+                pos = 0;
+                break;
+            }
+        }
+        runcamDeviceDispWriteChars(osdDevice, data, dataLen);
+
+        rcdeviceOSDLock = false;
+    }
+#endif
+    return 0;
+}
+
+int rcdeviceOSDWriteString(displayPort_t *displayPort, uint8_t x, uint8_t y, const char *buff)
+{
+    UNUSED(displayPort);
+#if !defined(USE_PARTICLE_DRAW)
+    runcamDeviceDispWriteHorizontalString(osdDevice, x, y, buff);
+#else
+    uint8_t i = 0;
+    for (i = 0; *(buff + i); i++) {
+        if (x + i < columnCount) { // Do not write over screen
+            screenBuffer[y * columnCount + x + i] = *(buff + i);
+        }
+    }
+#endif
+    return 0;
+}
+
+int rcdeviceOSDWriteChar(displayPort_t *displayPort, uint8_t x, uint8_t y, uint8_t c)
+{
+    UNUSED(displayPort);
+#if !defined(USE_PARTICLE_DRAW)
+    runcamDeviceDispWriteChar(osdDevice, x, y, c);
+#else
+    screenBuffer[y * columnCount + x] = c;
+#endif
+
+    return 0;
+}
+
+int rcdeviceOSDClearScreen(displayPort_t *displayPort)
+{
+    UNUSED(displayPort);
+
+#if defined(USE_PARTICLE_DRAW)
+    uint16_t x;
```

Comment:
why not `memset`? Compiler is probably clever enough to generate efficient (word access) code